algorithms/kernel/dtrees/gbt/gbt_train_tree_builder.i

```diff
@@ -274,22 +274,70 @@ protected:
         const size_t nSamples = _ctx.nSamples();
         const int* aSampleToF = _ctx.aSampleToF();
 
-        if(aSampleToF)
+        const size_t nThreads = _ctx.numAvailableThreads();
+        if ( nSamples < getThrOptBorder<cpu>(nThreads) || nThreads <= 1 )
         {
-            PRAGMA_VECTOR_ALWAYS
-            for(size_t i = 0; i < nSamples; ++i)
+            if ( aSampleToF )
             {
-                G += pgh[aSampleToF[i]].g;
-                H += pgh[aSampleToF[i]].h;
+                PRAGMA_VECTOR_ALWAYS
+                for (size_t i = 0; i < nSamples; ++i)
+                {
+                    G += pgh[aSampleToF[i]].g;
+                    H += pgh[aSampleToF[i]].h;
+                }
+            }
+            else
+            {
+                PRAGMA_VECTOR_ALWAYS
+                for (size_t i = 0; i < nSamples; ++i)
+                {
+                    G += pgh[i].g;
+                    H += pgh[i].h;
+                }
             }
         }
         else
         {
-            PRAGMA_VECTOR_ALWAYS
-            for(size_t i = 0; i < nSamples; ++i)
+            daal::services::internal::TArray<algorithmFPType, cpu> gsArr(nThreads);
+            daal::services::internal::TArray<algorithmFPType, cpu> hsArr(nThreads);
+            algorithmFPType * gs = gsArr.get();
+            algorithmFPType * hs = hsArr.get();
+            daal::services::internal::service_memset<algorithmFPType, cpu>(gs, 0, nThreads);
+            daal::services::internal::service_memset<algorithmFPType, cpu>(hs, 0, nThreads);
+            const size_t nPerBlock = nSamples / nThreads;
+            const size_t nSurplus = nSamples % nThreads;
+            if ( aSampleToF )
+            {
+                daal::threader_for(nThreads, nThreads, [&](size_t iBlock)
+                {
+                    size_t start = iBlock + 1 > nSurplus ? nPerBlock * iBlock + nSurplus : (nPerBlock + 1) * iBlock;
+                    size_t end = iBlock + 1 > nSurplus ? start + nPerBlock : start + (nPerBlock + 1);
+                    PRAGMA_VECTOR_ALWAYS
+                    for (size_t i = start; i < end; i++)
+                    {
+                        gs[iBlock] += pgh[aSampleToF[i]].g;
```

Comment:
It is really not efficient way to calculate sum by threads. I can expect, that scaling of this loop is not liner.
Due to "false sharing". it is an effect when you update one cache line by a few threads in the same time.

Alignment on 64 bytes should help